evaluator/evaluator.go

```diff
@@ -108,7 +108,9 @@ func Eval(node ast.Node, env *object.Environment) object.Object {
 		}
 
 		right := Eval(node.Right, env)
-		if isError(right) {
+		// If operator is "&&" or "||", we don't return right here to allow lazy evaluation
+		// right will be returned later in evalInfixExpression when lazy evlation doesn't apply
+		if isError(right) && !(node.Operator == "&&" || node.Operator == "||") {
```

Comment:
This code will still make it so that the right side is evaluated. Imagine we have something like:

```
true || 1..1000000000
```

This expression will take ages to evaluate even though it could theoretically return immediately.

What I'd suggest is to change `evalInfixExpression` to accept nodes rather than their evaluated value, and let it handle lazy evaluation on its own. Something like:

```
evalInfixExpression(operator, left, right) {
  leftVal = eval(left)

  if !isTruthy(leftVal) {
    if operator = && {
    return eval(right)
   }
    return false
  }

  if isTruthy(leftVal) && operator = || {
    return leftVal
  }

  rightVal = eval(right)

  switch:
    // all other cases
}
```

Reply:
I was thinking about this solution! This is more efficient (lazy) 👍